### PR TITLE
I1529 Frontend Dependancy Update

### LINKED ISSUE #1529: Upgrade to Frontend Dependencies
It's time we update the Myla Major FE libraries like React 8 and Material UI 5 we are one version behind. We are going to see a possible issue when Upgrading to ApolloClient as listed here https://github.com/tl-its-umich-edu/my-learning-analytics/issues/1525

----------
Test Plan
https://docs.google.com/spreadsheets/d/18c_VDE7wdme8c36Zo5-pjad77Kit8hFz4ylbmFXmLP8/edit#gid=0

### PR DESCRIPTION
**Pull request by pushyamig**
Fixes #1529 and #1525

1. Fix the UI changes due to MUI v5 update as it seems to be due to MUI adopting https://emotion.sh/docs/introduction
   1.  https://mui.com/material-ui/migration/migration-v4/. The documentation is nuanced since it talk all scenarios with upgrade to MUi v5 with React 17,  MUi v5 with React 18, MUi v5 supporting JSS etc
   2. used  CodeMods for updating the MUI to v5. here are the docs what gets updated https://github.com/mui/material-ui/blob/master/packages/mui-codemod/README.md#variant-prop
2. `react-router-dom` has some major changes, the nested router path are [relative now](https://reactrouter.com/en/main/upgrading/reach)
    1. `withRoute`r higher order hook is removed, so need to write a [custom function](https://reactrouter.com/en/main/start/faq#what-happened-to-withrouter-i-need-it)
3. React 18 had some changes to how [react is initiated](https://react.dev/blog/2022/03/08/react-18-upgrade-guide). lot of Improvement to async support.
4. The strategy for `package-lock.json` was just delete `node_module` and update the `package.json` and generate a new lock file. This was simplest!
5. upgrade to `@apollo/client`, Assignment planning was crashing so added a configuration option `assumeImmutableResults: false,` to enable mutable object of response which was allowed as an configuration option.
6. used `npm-check-updates` for checking the latest package version in the repo. It is recommended by package dev to install it globally. https://www.npmjs.com/package/npm-check-updates `npm install -g npm-check-updates
`
7.  [Three security advisory](https://github.com/tl-its-umich-edu/my-learning-analytics/security/dependabot) w.r.t (postcss, graphql, @babel/traverse) are resolved now. 
check using the command `npm ls <package-name>` Eg., `npm ls postcss`
              

  
## Todo List:
- [x] UI changes from MUI v5 upgrade ( There might be more just need to dig down) - **Jaydon**
     - [x] Landing page layout and View card layout
       - [x] disable button is not rendering color grey when the view is disabled and on hover
       - make sure to add Copy right flatpage (via django admin tool). I don't see anychanges needed here
     - [x] Resources access view tiny form
     - [x] Resource accessed Drop-down box is slight touching the slider. Needs some adjustment
     - [x] Making the bar %value visible with student accessed the file
     - [x] Assignment planning Pop-up window: no styling applied
          - [x] Assignment Planning pop-up: exiting mouse hover doesn't close the popover
     - [x] Survey Model is not centered after clicking the survey button. 
          - [x] Survey Model when opening up causing an warning in console log
     - [x] On hover of the survey button the the color contrast making it Opaque
           Reproduce it Admin -> Config -> SURVEY_URL
     - [x]  <ListItem button> the `Listitem ` component `button` attribute seems deprecated ( it is stricked out in VsCode). Needs an understanding if any impact and any otherway to compensate it. Documentation will help
     - Need an understanding in impact during the upgrade `import IconButton from '@mui/material/IconButton'` the iconButton Component added `size='large'` to all files
     - use of `sx` prop. MUI v5 introduced this prop for inline styling for certain component. for older version `style` prop was used. I don't feel we need to replace `style` with `sx` but building some understanding will help. https://mui.com/system/getting-started/the-sx-prop/
https://stackoverflow.com/questions/72527461/when-should-i-use-style-instead-of-sx-prop-in-material-ui#:~:text=sx%20prop%20works%20only%20on,some%20cases%20as%20explained%20below.
- [x] Webpack/babel lib update and build update - **Pushyami**
- [x] Other packages update react-router-dom, rc-slider, d3 etc - **Pushyami**
- [x] strategy around `package-lock.json` commit - **Pushyami**
- [ ] Over all testing - **Pushyami/Jaydon**
- [ ] ~~Jest testing as things are not working as expected~~ (https://github.com/tl-its-umich-edu/my-learning-analytics/issues/1558)

---------
## Test Plan
https://docs.google.com/spreadsheets/d/18c_VDE7wdme8c36Zo5-pjad77Kit8hFz4ylbmFXmLP8/edit#gid=0

`Testing should be done from Student, instructor and MyLa Admin role`

### CODE REVIEW

File: assets/src/containers/WithRouter.js [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
+      />
     )
   }
 
   return ComponentWithRouterProp
 }
-export default withRouter
\ No newline at end of file
+
+export default withRouter
```
Comment (pushyamig) [inline assets/src/containers/WithRouter.js:24]: V6 of the react-router-dom removed withRouter, so I have to write this HOC as per the documentation guide, linked in the page

File: assets/src/containers/App.js [hunk 2]
```diff
@@ -18,15 +18,17 @@ function App (props) {
...
-  
   return (
     <>
       <Helmet titleTemplate='%s | My Learning Analytics' title='Courses' />
       <GoogleAnalyticsTracking {...{ gaId, cspNonce }} />
       <Routes>
         <Route path='/' element={<CourseList user={user} />} />
-        <Route path='/courses' element={ <CourseList user={user} />} />
-        <Route path='/courses/:courseId/*' element={courseId ? <Course user={user} courseId={Number(courseId)} {...props} /> : <AlertBanner>Application Launch did not happened properely</AlertBanner>} />
+        <Route path='/courses' element={<CourseList user={user} />} />
```
Comment (pushyamig) [inline assets/src/containers/App.js:27]: part of react-router-dom v6 change

File: assets/src/components/d3/createResourceAccessChart.js [hunk 3]
```diff
@@ -256,8 +256,8 @@ function createResourceAccessChart ({ data, width, height, domElement }) {
     const fullRange = mainYZoom.range()
     const selection = event
       ? event.selection[1] === 0 // prevents [0, 0] from being returned, which causes bug
-          ? [0, 0.1]
-          : event.selection
+        ? [0, 0.1]
+        : event.selection
       : defaultSelection
```
Comment (pushyamig) [inline assets/src/components/d3/createResourceAccessChart.js:261]: Spacing issue

File: assets/src/components/RangeSlider.js [hunk 4]
```diff
@@ -35,15 +35,17 @@ const rangeSlider = props => {
         range
         activeDotStyle={activeDotStyle}
         dotStyle={dotStyle}
-        handleStyle={[activeDotStyle, activeDotStyle]}
-        railStyle={unselectedStyle}
-        trackStyle={[selectedStyle]}
```
Comment (pushyamig) [inline assets/src/components/RangeSlider.js:39]: These are deprecated in the new version of rc-slider and includes as part of the styles property.

File: assets/src/components/RangeSlider.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import 'rc-slider/assets/index.css'
-import Slider, { Range } from 'rc-slider'
```
Comment (pushyamig) [inline assets/src/components/RangeSlider.js]: rc-slider package has some solid change, but not breaking. I felt the slider looked ok and functioned as before

File: assets/src/components/ProgressBarV2.js [hunk 6]
```diff
@@ -65,11 +65,11 @@ function ProgressBarV2 (props) {
...
-                    <div key={key}>
-                      <Line {...line} barHeight={height} outOf={outOf} />
-                    </div>
-                  ))
+                .map((line, key) => (
+                  <div key={key}>
+                    <Line {...line} barHeight={height} outOf={outOf} />
+                  </div>
+                ))
               : null
```
Comment (pushyamig) [inline assets/src/components/ProgressBarV2.js]: Spacing change

File: assets/src/components/SelectCard.js [hunk 7]
```diff
@@ -235,11 +235,11 @@ const SelectCard = props => {
       </CardContent>
     )
 
-    return <Root>{cardImage}{cardContent}</Root>
+    return <>{cardImage}{cardContent}</>
   }
```
Comment (pushyamig) [inline assets/src/components/SelectCard.js:239]: This is nice and simple fix. I pulled the latest and things are looking good with this commit.

File: assets/src/components/SurveyModal.js [hunk 8]
```diff
@@ -92,7 +92,7 @@ export default function SurveyModal (props) {
   const toggleOpen = () => setOpen(!open)
 
   const body = (
-    <Root className={`${classes.paper} ${classes.modal}`}>
```
Comment (pushyamig) [inline assets/src/components/SurveyModal.js]: @jaydonkrooss I couldn't test this, but I felt this would work. were you able to test this with survey link?

Comment (pushyamig) [reply assets/src/components/SurveyModal.js]: The button looks good, But the modal after clicking is not centered. I added 2 more issue around this. 

You can reproduced it locally by going to Admin -> Config -> SURVEY_URL "Give any URL"

**Comment by pushyamig** [comment]
I have built the latest from this repo which is here https://github.com/pushyamig/my-learning-analytics/actions/runs/7413997544

**Comment by pushyamig** [comment]
This PR is now ready for Review, I have added test plan as well

**Comment by pushyamig** [comment]
I pretty much tested/reviewed it from my lense and things are ok .

I have updated things I tested according to test plan
https://docs.google.com/spreadsheets/d/18c_VDE7wdme8c36Zo5-pjad77Kit8hFz4ylbmFXmLP8/edit#gid=0

**Review by jaydonkrooss — APPROVED** [review]

**Comment by pushyamig** [comment]
I can't approve a PR which is opened by me. But I will approve the Jaydon's work as well

**Pull request merged**